src/distance/osa.rs

```diff
@@ -421,129 +500,208 @@ impl<CharT> MetricUsize for BatchComparator<CharT> {
     }
 }
 
-impl<Elem1> BatchComparator<Elem1>
-where
-    Elem1: HashableChar + Clone,
+impl<'a, Elem1, MetricType, ResultType>
+    BatchComparatorBuilder<'a, Elem1, MetricType, ResultType, NoScoreCutoff>
 {
-    pub fn new<Iter1>(s1_: Iter1) -> Self
+    fn new(cache: &'a BatchComparator<Elem1>) -> Self {
+        Self {
+            cache,
+            score_cutoff: Default::default(),
+            score_hint: Default::default(),
+            metric_type: PhantomData,
+        }
+    }
+}
+
+impl<'a, Elem1, MetricType, ResultType, CutoffType>
+    BatchComparatorBuilder<'a, Elem1, MetricType, ResultType, CutoffType>
+{
+    pub fn with_score_hint(mut self, score_hint: impl Into<ResultType>) -> Self {
+        self.score_hint = Some(score_hint.into());
+        self
+    }
+
+    pub fn with_score_cutoff(
+        self,
+        score_cutoff: impl Into<ResultType>,
+    ) -> BatchComparatorBuilder<'a, Elem1, MetricType, ResultType, WithScoreCutoff<ResultType>>
+    {
+        BatchComparatorBuilder {
+            cache: self.cache,
+            score_hint: self.score_hint,
+            score_cutoff: WithScoreCutoff(score_cutoff.into()),
+            metric_type: self.metric_type,
+        }
+    }
+}
+
+impl<Elem1> BatchComparatorBuilder<'_, Elem1, Dist, usize, NoScoreCutoff> {
+    pub fn compare<Iter2>(&self, s2: Iter2) -> usize
     where
-        Iter1: IntoIterator<Item = Elem1>,
-        Iter1::IntoIter: Clone,
+        Iter2: IntoIterator,
+        Iter2::IntoIter: DoubleEndedIterator + Clone,
+        Elem1: PartialEq<Iter2::Item> + HashableChar + Copy,
+        Iter2::Item: PartialEq<Elem1> + HashableChar + Copy,
     {
-        let s1_iter = s1_.into_iter();
-        let s1: Vec<Elem1> = s1_iter.clone().collect();
+        // todo this calls s1.count internally is this optimizes in the same way it
+        // would if we pass in self.cache.s1.len()
```

Comment:
I didn't compare the assembler, but performance wise it doesn't make any noticeable difference. Would have been surprising too, since this iterator should know it's size.